oap-server/analyzer/agent-analyzer/src/main/java/org/apache/skywalking/oap/server/analyzer/provider/trace/TraceSamplingPolicyWatcher.java

```diff
@@ -0,0 +1,263 @@
+            this.settingsString.set(ymlContent);
+            return setting;
+        } catch (Exception e) {
+            log.error("[trace-sample-policy] Failed to parse yml content: \n{}", ymlContent, e);
+        }
+        return null;
+    }
+
+    private SampleRateSetting mapToSettingObject(Map<String, Object> map) {
+        if (map == null) {
+            return null;
+        }
+        SampleRateSetting setting = new SampleRateSetting();
+        DefaultSampleRateConfig defaultSampleRateConfig = new DefaultSampleRateConfig();
+        // default
+        Object defaultMapObject = map.get("default");
+        if (defaultMapObject != null) {
+            Map<String, Object> defaultMap = (Map<String, Object>) defaultMapObject;
+            defaultSampleRateConfig.setSampleRate(new AtomicReference<Integer>((Integer) defaultMap.getOrDefault("sampleRate", null)));
+            defaultSampleRateConfig.setDuration(new AtomicReference<Integer>((Integer) defaultMap.getOrDefault("duration", null)));
+        }
+        setting.setDefaults(defaultSampleRateConfig);
+        // services
+        List<ServiceSampleConfig> services = new ArrayList<>();
+        Object servicesObject = map.get("services");
+        if (servicesObject != null) {
+            List<Map<String, Object>> serviceList = (List<Map<String, Object>>) servicesObject;
+            serviceList.forEach(service -> {
+                ServiceSampleConfig serviceSampleConfig = new ServiceSampleConfig();
+                serviceSampleConfig.setName((String) service.get("name"));
+                serviceSampleConfig.setSampleRate(service.get("rate") == null ? null : new AtomicReference<Integer>((Integer) service.get("rate")));
+                serviceSampleConfig.setDuration(service.get("duration") == null ? null : new AtomicReference<Integer>((Integer) service.get("duration")));
+                services.add(serviceSampleConfig);
+            });
+        }
+        setting.setServices(services);
+        return setting;
+    }
+
+    @Getter
+    @Setter
+    @ToString
+    public static class ServiceSampleConfig {
+        private String name;
+        private AtomicReference<Integer> sampleRate;
+        private AtomicReference<Integer> duration;
+    }
+
+    @ToString
+    public static class SampleRateSetting implements Iterable<ServiceSampleConfig> {
+        @Getter
+        @Setter
+        private DefaultSampleRateConfig defaults;
+        @Getter
+        @Setter
+        private Collection<ServiceSampleConfig> services;
+
+        SampleRateSetting() {
+            services = new ArrayList<>();
+        }
+
+        @Override
+        public Iterator<ServiceSampleConfig> iterator() {
+            return services.iterator();
+        }
+    }
+
+    @Getter
+    @Setter
+    @ToString
+    public static class DefaultSampleRateConfig {
+        private AtomicReference<Integer> sampleRate;
+        private AtomicReference<Integer> duration;
```

Comment:
No need to use `AtomicXXX`, primitives should be enough

Reply:
And it seems we can make `DefaultSampleRateConfig` immutable, so that no thread safety problem